Sources/Controllers/2.0/OASelectSubcategoryViewController.mm

```diff
@@ -7,214 +7,285 @@
-        }];
-
-        _data = [NSArray arrayWithArray:data];
-        _keys = [NSArray arrayWithArray:keys];
+        NSSet<NSString *> *acceptedTypes = [[_filter getAcceptedTypes] objectForKey:_category];
+        NSSet<NSString *> *acceptedSubtypes = [_filter getAcceptedSubtypes:_category];
+        NSArray<OAPOIType *> *types = _category.poiTypes;
+
+        _selectedItems = [NSMutableArray new];
+        _items = [NSMutableArray arrayWithArray:[types sortedArrayUsingComparator:^NSComparisonResult(OAPOIType * _Nonnull t1, OAPOIType * _Nonnull t2) {
+            return [t1.nameLocalized localizedCaseInsensitiveCompare:t2.nameLocalized];
+        }]];
+
+        if (acceptedSubtypes == [OAPOIBaseType nullSet] || acceptedTypes.count == types.count)
+            _selectedItems = [NSMutableArray arrayWithArray:_items];
+        else
+            for (OAPOIType *poiType in _items)
+                if ([acceptedTypes containsObject:poiType.name])
+                    [_selectedItems addObject:poiType];
+
+        [_items insertObject:[[OACustomSelectionButtonCell alloc] init] atIndex:0];
     }
 }
 
 - (void)viewDidLoad
 {
     [super viewDidLoad];
-    
+
     if (_category)
-        self.textView.text = _category.nameLocalized;
+        self.titleLabel.text = _category.nameLocalized;
     else
-        self.textView.text = @"";
-    
-    // add header
-    _headerView = [[OAMultiselectableHeaderView alloc] initWithFrame:CGRectMake(0.0, 1.0, 100.0, 44.0)];
-    [_headerView setTitleText:OALocalizedString(@"select_all")];
-    
-    _headerView.section = 0;
-    _headerView.delegate = self;
-    
-    self.tableView.editing = YES;
-    if (_selectAll)
-    {
-        _headerView.selected = YES;
-        [self.tableView beginUpdates];
-        for (int i = 0; i < _data.count; i++)
-            [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0] animated:NO scrollPosition:UITableViewScrollPositionNone];
-        [self.tableView endUpdates];
-    }
-    else
-    {
-        _headerView.selected = _subcategories.count == _keys.count;
-        [self.tableView beginUpdates];
-        for (int i = 0; i < _keys.count; i++)
-        {
-            NSString *name = _keys[i];
-            if ([_subcategories containsObject:name])
-                [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0] animated:NO scrollPosition:UITableViewScrollPositionNone];
-        }
-        [self.tableView endUpdates];
-    }
-    [self applySafeAreaMargins];
-}
+        self.titleLabel.text = @"";
 
--(UIView *) getTopView
-{
-    return _topView;
-}
+    self.tableView.delegate = self;
+    self.tableView.dataSource = self;
+    self.tableView.editing = YES;
+    self.tableView.tintColor = UIColorFromRGB(color_primary_purple);
 
--(UIView *) getMiddleView
-{
-    return _tableView;
+    [self.tableView beginUpdates];
+    for (int i = 1; i < _items.count; i++)
```

Comment:
Prefer using NSInteger in Objective C code, use int only if required by the backend implementation (e.g. direct code port from Java).